### ISSUE TITLE
OpenCL missing

### ISSUE THREAD

**Issue by knight-ryu12** [zoogie/seedminer#2]
therefore (ref #1), OpenCL is licensed DLL and card-dependent.
NVIDIA's OpenCL lib can be obtained at their homepage.

**Comment by zoogie** [comment]
openCL is supposed to be platform independent.

What exactly happens when you try to run bfCL?

**Comment by knight-ryu12** [comment]
OpenCL.dll are missing.


2018/02/04 午前8:07 "zoogie" <notifications@github.com>:

**Comment by zoogie** [comment]
Are you running an old windows or something? Mine's in the system32 dir which means it's part of windows.
Nobody else seems to be running into this problem.

**Comment by Jimmy-Z** [comment]
opencl.dll should be provided by the system, never be shipped with the app, it's highly likely your system is corrupted, or just too old.

And it is provided by Khronos group(the standardization entity behind OpenCL) and not _card_ dependent, state your problem, not your assumptions.

**Comment by knight-ryu12** [comment]
NVIDIA CUDA Toolkit worked.
My system is Windows 10 Home TP, is my system so old for that? I don't
think so.

2018/02/04 午後11:37 "Jimmy-Z" <notifications@github.com>:

**Comment by knight-ryu12** [comment]
maybe it was my system… that I am not sure about it.


2018/02/05 午前0:11 "Dragon ryu" <knight.ryu12@gmail.com>:

**Comment by Jimmy-Z** [comment]
CUDA is not OpenCL.

**Comment by knight-ryu12** [comment]
Then why it worked?
I am sure that toolkit did provide me openCL.DLL

2018/02/05 午前0:22 "Jimmy-Z" <notifications@github.com>:

**Comment by Jimmy-Z** [comment]
I've no idea why nvidia bundles opencl.dll with "cuda toolkit", ask nvidia.

I thought you meant "nvidia cuda toolkit worked (so my system is fine)", but from your next reply it seems to be "after I installed nvidia cuda toolkit, bfCL works now".

At least 3 of my friends with nvidia GPU tested bfCL without "nvidia cuda toolkit".

**Comment by angelsl** [comment]
> opencl.dll should be provided by the system

OpenCL is certainly not provided by Windows. You need to install some OpenCL provider (which includes NVIDIA GPUs, of course).

1. [NVIDIA CUDA Toolkit](https://developer.nvidia.com/cuda-toolkit) (it seems NVIDIA drivers include the OpenCL driver too, but I'm not sure)
2. [AMD OpenCL Driver](http://support.amd.com/en-us/kb-articles/Pages/OpenCL2-Driver.aspx)
3. [Intel OpenCL Driver](https://software.intel.com/en-us/articles/opencl-drivers) (yes, you can even use your Intel CPU proper, not just the integrated graphics, as an OpenCL provider, as useful as that is)

**Comment by knight-ryu12** [comment]
1. openCL is not provided with Windows. need GPU dependent package.
2. bfCL, yes. Sorry.
3. ASk theirs, as windows (vanila) clearly doesn't have openCL.

**Comment by hax0kartik** [comment]
Can confirm vanilla windows doesn't have OpenCl.dll

**Comment by angelsl** [comment]
FWIW, @zoogie 

> Mine's in the system32 dir which means it's part of windows.

Many, many programs install DLLs into System32.

**Comment by Jimmy-Z** [comment]
Vendor specific OpenCL implementations is installed as ICD and loaded by the vendor independent opencl.dll.

![image](https://user-images.githubusercontent.com/85773/35779216-5196f0ca-0a04-11e8-815f-c88dcf43a528.png)

By system I meant the underlying system, either by Windows or GPU driver, I said "never be shipped with the app", read context.

**Comment by knight-ryu12** [comment]
Then add some diarect?
Clearly all "common users" trying to use this seedminer will face this issue.

**Comment by Jimmy-Z** [comment]
I'm quite sure your case can't present "all common users"

**Comment by knight-ryu12** [comment]
I'm quite sure that vanilla windows, doesn't come with openCL.
Failing to launch bfCL will result not working (No Error message shown.)
adding dialect isn't hard to do.
it's up to you about this issue, weather close it or not.

**Comment by Jimmy-Z** [comment]
"never be shipped with the app", this is the second time I quote this.

"No error message shown"? where did you get that "opencl.dll are missing"?

**Comment by Jimmy-Z** [comment]
https://forums.khronos.org/showthread.php/11392-Yet-another-discussion-about-OpenCL-dll

Some related discussion.

**Comment by knight-ryu12** [comment]
When I said ship openCL with bfCL?
I am just complaining about there were no dialect with openCL.  In my
humble opinion, I never thought that this seedminer uses openCL to
accomplish gpu datamining.

Second, python script doesn't show error at exes no matter what.
I just double clicked exes itself.


2018/02/05 午前0:59 "Jimmy-Z" <notifications@github.com>:

**Comment by Jimmy-Z** [comment]
OK I assumed that's your proposed solution.

To be clear I'm the author of bfCL, @zoogie is the author of seedminer. seedminer calls bfCL.

You should understand your case is not common, I'm not sure if vanilla windows comes with opencl.dll, but I've tested on 4 of my Windows 10 computers(1 AMD + Intel, 1 AMD only, 2 Intel only), and several friends(various including NVIDIA) helped me testing, none of them complained this issue. If former statement of "vanilla windows doesn't have opencl.dll" is true, I'd assume it comes with GPU driver, and supposedly you have that installed, either way it's likely your system is corrupted.

**Comment by Jimmy-Z** [comment]
By the way, opencl.dll doesn't equal OpenCL.

**Comment by Jimmy-Z** [comment]
> OpenCL is certainly not provided by Windows. You need to install some OpenCL provider (which includes NVIDIA GPUs, of course).
NVIDIA CUDA Toolkit (it seems NVIDIA drivers include the OpenCL driver too, but I'm not sure)
AMD OpenCL Driver
Intel OpenCL Driver (yes, you can even use your Intel CPU proper, not just the integrated graphics, as an OpenCL provider, as useful as that is)

@angelsl 
Open those links, the AMD OpenCL Driver page are just _old_ links to their GPU driver. and I quote the Intel page: `(Also automatically shipped with Windows graphics drivers)`

In short: all major GPU vendors include OpenCL in their GPU driver packages.

And this issue is *NOT* related.

**Comment by Jimmy-Z** [comment]
@Pirater12 I don't know what's your definition of "vanilla windows", but I just mounted a wim image from a Windows 10 1709 iso downloaded from Microsoft, and it does include opencl.dll.

It appears in 4 different places in Pro and 2 different places in Home.

**Comment by angelsl** [comment]
> I don't know what's your definition of "vanilla windows", but I just mounted a wim image from a Windows 10 1709 iso downloaded from Microsoft, and it does include opencl.dll.
>
> It appears in 4 different places in Pro and 2 different places in Home.

As far as I can tell OpenCL.dll is present in WinSxS but only active in SysWOW64 i.e. for 32-bit applications. A 64-bit OpenCL.dll exists in WinSxS but is not active, and needs to be installed separately.

> all major GPU vendors include OpenCL in their GPU driver packages.

I've had this exact issue (missing OpenCL.dll) in the past when using OpenCL cryptocurrency miners without having installed NVIDIA's CUDA toolkit. (I had NVIDIA drivers installed, of course.) So, maybe Chromaryu and I were just unlucky, but it certainly is not unusual for OpenCL.dll to not be present.

**Comment by Jimmy-Z** [comment]
OK, all major GPU vendors include OpenCL in their GPU driver packages _now_.
